nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py

```diff
@@ -1172,14 +1172,17 @@ def loss_func(self, loss_mask, num_valid_tokens_in_ub, output_tensor):
         return loss
 
     def build_train_valid_test_datasets(self):
-        # Override limit_val_batches to be a multiple of num microbatches to prevent val_step from exiting in between a step
-        self._reconfigure_val_batches()
-        logging.info('Building GPT datasets.')
         if self.trainer.limit_val_batches > 1.0 and isinstance(self.trainer.limit_val_batches, float):
             raise ValueError("limit_val_batches must be an integer or float less than or equal to 1.0.")
+        logging.info('Building GPT datasets.')
         global_batch_size = self.cfg.global_batch_size
         max_train_steps = self.trainer.max_steps
-        eval_iters = (max_train_steps // self.trainer.val_check_interval + 1) * self.trainer.limit_val_batches
+        # if limit_val_batches is 0, don't use it for computing eval samples, as it can cause error in building the dataset with 0 samples
```

Comment:
Building dataset with 0 samples causes this error: `index -1 is out of bounds for axis 0 with size 0` in this [line](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/datasets/blended_dataset.py#L92) in blended_dataset.py in mcore. I agree with you that the dataset and the dataloaders shouldn't be constructed if the num_samples is 0. Since the dataset building happens in mcore, this needs to be done there. This can't be done in NeMo since we have one func call to build all 3 of train, val and test datasets [here](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L1232)

Reply:
For this error `index -1 is out of bounds for axis 0 with size 0` mentioned above, the index becomes -1 in this [line](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/datasets/blended_dataset.py#L81) since `self.size = 0` (if `limit_val_batches=0`)

Reply:
Let me check in mcore if we can make a PR, to add a condition check to not build dataset if the `size` or `num_samples` is 0.

Reply:
Can you post the full stack trace for the error you get?

Reply:
```
Exception has occurred: IndexError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
index -1 is out of bounds for axis 0 with size 0
  File "/usr/local/lib/python3.10/dist-packages/megatron/core/datasets/blended_dataset.py", line 92, in __getitem__
    dataset_id = self.dataset_index[idx]
  File "/usr/local/lib/python3.10/dist-packages/megatron/core/datasets/blended_dataset.py", line 81, in __init__
    _ = self[self.size - 1]
  File "/usr/local/lib/python3.10/dist-packages/megatron/core/datasets/blended_megatron_dataset_builder.py", line 276, in build_generic_dataset
    dataset = cls(*args)
  File "/usr/local/lib/python3.10/dist-packages/megatron/core/datasets/blended_megatron_dataset_builder.py", line 117, in _build_blended_dataset_splits
    self.build_generic_dataset(
  File "/usr/local/lib/python3.10/dist-packages/megatron/core/datasets/blended_megatron_dataset_builder.py", line 61, in build
    return self._build_blended_dataset_splits()
  File "/workspace/software/NeMo/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py", line 1235, in build_train_valid_test_datasets
    ).build()
  File "/workspace/software/NeMo/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py", line 1325, in setup
    self.build_train_valid_test_datasets()
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/call.py", line 145, in _call_lightning_module_hook
    output = fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/call.py", line 86, in _call_setup_hook
    _call_lightning_module_hook(trainer, "setup", stage=fn)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/trainer.py", line 941, in _run
    call._call_setup_hook(self)  # allow user to setup lightning_module in accelerator environment
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/trainer.py", line 571, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/launchers/subprocess_script.py", line 93, in launch
    return function(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/call.py", line 42, in _call_and_handle_interrupt
    return trainer.strategy.launcher.launch(trainer_fn, *args, trainer=trainer, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/trainer.py", line 532, in fit
    call._call_and_handle_interrupt(
  File "/workspace/software/NeMo/examples/nlp/language_modeling/megatron_gpt_pretraining.py", line 38, in main
    trainer.fit(model)
  File "/usr/local/lib/python3.10/dist-packages/hydra/core/utils.py", line 186, in run_job
      ret.return_value = task_function(task_cfg)
  File "/usr/local/lib/python3.10/dist-packages/hydra/core/utils.py", line 260, in return_value
    raise self._return_value
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/hydra.py", line 132, in run
    _ = ret.return_value
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/utils.py", line 458, in <lambda>
    lambda: hydra.run(
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/utils.py", line 223, in run_and_report
    raise ex
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/utils.py", line 223, in run_and_report
    raise ex
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/utils.py", line 457, in _run_app
    run_and_report(
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/utils.py", line 394, in _run_hydra
    _run_app(
  File "/workspace/software/NeMo/nemo/core/config/hydra_runner.py", line 129, in wrapper
    _run_hydra(
  File "/workspace/software/NeMo/examples/nlp/language_modeling/megatron_gpt_pretraining.py", line 42, in <module>
    main()
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main (Current frame)
    return _run_code(code, main_globals, None,
IndexError: index -1 is out of bounds for axis 0 with size 0
```

Reply:
It errors out at this line [blended_dataset.py#L92](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/datasets/blended_dataset.py#L92) with the above error.